fix(linter/plugins): set up global scope correctly

overlookmotel opened the pull request:
Overhaul how variables in the global scope are set up.

* Create globals according to `env` and `globals` from config - previously we always used a standard "esnext" set.
* Remove the patch to `@typescript-eslint/scope-manager`. Instead do what ESLint does and patch up the global scope programmatically after TS-ESLint has produced the scope tree.
* Set `writable` property for global variables according to config.

This still isn't quite correct. There are many subtleties to this and I haven't got to the bottom of it all. But this PR fixes 500 conformance tests, so it's a large step in the right direction.

I haven't added more tests in this PR, because I'm not actually sure what behavior is correct, and what still needs to be fixed. Once all the conformance tests related to globals (which is most of the remaining failing ones) are fixed, we'll have the full picture and can add more tests.

overlookmotel commented on apps/oxlint/src-js/plugins/scope.ts:
```diff
@@ -113,10 +117,120 @@ function initTsScopeManager() {
+  // typescript-eslint uses `leftToBeResolved`, eslint-scope uses `left`
+  implicit.leftToBeResolved = implicit.leftToBeResolved.filter(
+    (ref: Reference) => !globalScope.set.has(ref.identifier.name),
+  );
```
No. An error is better than silent malfunction. If it doesn't exist, our code is wrong and we need to know about it.

overlookmotel commented on apps/oxlint/conformance/src/rule_tester.ts:
```diff
@@ -93,19 +106,96 @@ function modifyTestCase(test: TestCase): void {
+        return eslintGlobals[`es${ecmaVersion + 2009}`];
+      }
+      // es2015 and later use the year directly
+      return eslintGlobals[`es${ecmaVersion}`];
```
This in the conformance tests. The whole point is to produce errors if things aren't correct.

camc314 commented:
### Merge activity

* **Dec 23, 10:53 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Dec 23, 10:55 AM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
fix(linter/plugins): set up global scope correctly (#17293)

Overhaul how variables in the global scope are set up.

* Create globals according to `env` and `globals` from config - previously we always used a standard "esnext" set.
* Remove the patch to `@typescript-eslint/scope-manager`. Instead do what ESLint does and patch up the global scope programmatically after TS-ESLint has produced the scope tree.
* Set `writable` property for global variables according to config.

This still isn't quite correct. There are many subtleties to this and I haven't got to the bottom of it all. But this PR fixes 500 conformance tests, so it's a large step in the right direction.

I haven't added more tests in this PR, because I'm not actually sure what behavior is correct, and what still needs to be fixed. Once all the conformance tests related to globals (which is most of the remaining failing ones) are fixed, we'll have the full picture and can add more tests.
